### ISSUE TITLE
Label and RichTextLabel breaks CJK line-break rules (禁則処理 / 줄 바꿈 규칙)

### ISSUE THREAD

**Issue by JapanYoshi** [godotengine/godot#65636]
### Godot version

3.5

### System information

Windows 10, GLES3, Vulkan

### Issue description

I want to make a Japanese game with Godot, but the included Label elements break Japanese typography rules. (Similar rules exist in Chinese and Korean text.)

For example, here is a diagram of a text box with improper line endings, with "〼" used to show the bounding box of the text box:

    〼〼〼〼〼〼〼〼〼〼〼
    〼ゴドー・エンジンは〼
    〼、だれでも無償で使〼
    〼えるゲームエンジン〼
    〼。　　　　　　　　〼
    〼〼〼〼〼〼〼〼〼〼〼

(By the way, the text translates to “Godot Engine is a game engine that anyone can use for free.”)

In this example, line 2 and line 4 start with a forbidden character: a comma and a period respectively. It should really look like this:

    〼〼〼〼〼〼〼〼〼〼〼
    〼ゴドー・エンジン　〼 < Line shortened by 1 character
    〼は、だれでも無償で〼
    〼使えるゲームエンジ〼
    〼ン。　　　　　　　〼
    〼〼〼〼〼〼〼〼〼〼〼

The following list of characters is from [the Japanese Wikipedia page](https://ja.wikipedia.org/wiki/%E7%A6%81%E5%89%87%E5%87%A6%E7%90%86) for “Line breaking rules in East Asian languages”.

The following letters cannot be at the start of a line, and the previous letter should be brought with it into a new line:

* Punctuation: 、。，．：；！？?!‼⁇⁈⁉
* Closing quotation marks and parentheses: ,)]｝、〕〉》」』】〙〗〟’”｠»
* Hyphens: ‐゠–〜～
* Letters that link to the previous letter: ゝゞーァィゥェォッャュョヮヵヶぁぃぅぇぉっゃゅょゎゕゖㇰㇱㇲㇳㇴㇵㇶㇷㇸㇹㇷ゚ㇺㇻㇼㇽㇾㇿ々〻

The following letters cannot be at the end of a line, and it should be placed with the next character on a new line:

* Opening quotation marks and parentheses: ([｛〔〈《「『【〘〖〝‘“｟«

The following pairs of letters should not be separated:

* ——
* ……
* ‥‥
* 〳〵
* 〴〵

### Steps to reproduce

1. Create a project that includes a Japanese font and DynamicFont
2. Create a scene containing a Label with line wrap enabled, and/or a RichTextLabel
3. Give the Label some text including these forbidden characters

### Minimal reproduction project

[cjk_line_break.zip](https://github.com/godotengine/godot/files/9541692/cjk_line_break.zip)

**Comment by gotnospirit** [comment]
Wrapping is basically just a way to say if it's ok (or not) to cut words in two when `line.width > container.width`, there's no grammar rules involved. I guess manually insert new lines is the only way to go

**Comment by Chaosus** [comment]
cc @bruvzg

**Comment by JapanYoshi** [comment]
> Wrapping is basically just a way to say if it's ok (or not) to cut words in two when `line.width > container.width`, there's no grammar rules involved. I guess manually insert new lines is the only way to go

The current behavior is like if words didn't stick together in Engish text. It's as acceptable for Japanese text as if this happened for English text:

    Godot Engine i
    s a free and o
    pen-source gam
    e engine.

You can see why manually inserting newlines is not a viable solution.

**Comment by Cykyrios** [comment]
Would you mind trying Godot 4 for this? I know this won't be a solution for Godot 3.x, but the new TextServer handles this kind of thing much better. On the other hand, it seems to always keep a constant number of characters for every line, so I get the following behavior when setting autowrap to "word":
```
〼〼〼〼〼〼〼〼〼〼〼
〼ゴドー・エンジン　〼
〼は、だれでも無償　〼
〼で使えるゲームエ　〼
〼ンジン。　　　　　〼
〼〼〼〼〼〼〼〼〼〼〼
```

**Comment by JapanYoshi** [comment]
> Would you mind trying Godot 4 for this? I know this won't be a solution for Godot 3.x, but the new TextServer handles this kind of thing much better.

I’m not sure where to get access to Godot 4; I thought it was still in alpha?

> On the other hand, it seems to always keep a constant number of characters for every line, 

That happens more often in Japanese fonts because most characters take 1 em of width, but if you mix plain Latin characters in it (e.g. replacing ゴドー・エンジン with “Godot Engine”), you’d still need to calculate line width, since they’re all different widths in proportional fonts, and 0.5 em wide in fixed-width fonts.

**Comment by voylin** [comment]
> I’m not sure where to get access to Godot 4; I thought it was still in alpha?

https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-16
Here you can get the most recent Alpha, just go to the download section for this.

Regarding this issue, it would be practically impossible to do this for Japanese as all characters are connected to each other. I have some experience in using Japanese inside of the engine and there are a couple of things that don't seem to work that efficiently when using Japanese. This is one of them.

There are 2 solutions for this which I could advise:
- 1: Make your own script that enters the new lines where needed. This would take a lot of work
- 2: Use zero-width spaces after each word. (Alt+255 or Alt+0160), those are known as non-breaking spaces and that's what I'm using right now for a game I'm making which is in Japanese.

Hope this helps.

**Comment by Cykyrios** [comment]
Yes, Godot 4 is still in alpha, you can get a build from [here](https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-16#downloads).
I just tried writing some words with hiragana only and can confirm that Labels do not follow the line breaking rules in Godot 4 either, as combinations such as ちゃ, にゅ and such can result in ゃ or ゅ being the first character of a new line.

**Comment by bruvzg** [comment]
> I just tried writing some words with hiragana only and can confirm that Labels do not follow the line breaking rules in Godot 4 either

Godot 4 should follow the ICU breaking rules when autowrap mode is set `Word` (`Arbitrary` is intentionally breaking without any rules, `Word (Smart)` will attempt to break words, if it's possible and then fallback to `Arbitrary`).

https://user-images.githubusercontent.com/7645683/193519232-3b628514-e104-4507-9c66-dd63948302f9.mov

**Comment by Cykyrios** [comment]
OK, I had to make sure I wasn't going crazy as I got both results while testing... Setting wrapping to either Word or Word (Smart) seems to be working as expected, _except_ if I set the language to ja_JP in the BiDi category, in which case I get the following, where 。 seems to be handled properly, but some lines start with ゃ or っ.

https://user-images.githubusercontent.com/53737317/193650952-f94a7009-b278-4fe8-8660-65cb57f7a5c9.mp4

**Comment by bruvzg** [comment]
Seems like it wants `lb=strict` modifier to prevent this, like: `ja_JP@lb=strict`. Not sure why it assumes it when language in the BiDi is not set (in this case it's using UI language code).

See https://www.w3.org/TR/css-text-3/#propdef-line-break

**Comment by BlooRabbit** [comment]
I used a little hack in Godot 3.5, by trying to list each of the relevant unicodes in Japanese and making them "unseperatable". It seems to work in my use cases: `						// Japanese non separatable symbols
						int const japchars [60]= {0x3002,0x3001,0xff0c,0xff0e,0xff1a,0xff1b,0xff01,0xff1f,0x309c,0x301c,0x309d,0x309e,
							0x30fc,0x30a1,0x30a3,0x30a5,0x30a7,0x30a9,0x30c3,0x30e3,0x30e5,0x30e7,0x30ee,0x30f5,0x30f6,0x3041,0x3043,
							0x3045,0x3047,0x3049,0x3063,0x3083,0x3085,0x3087,0x308e,0x30a3,0x3096,0x31f0,0x31f1,0x31f2,0x31f3,0x31f4,
							0x31f5,0x31f6,0x31f7,0x31f8,0x31f9,0x31f7,0x309a,0x31fa,0x31fb,0x31fc,0x31fd,0x31fe,0x31ff,0x3005,0x303b,
							0x2010,0x30a0,0x2013 };
						int n=0;
						for (n = 0; n < 2; ++n)
						{
							if (c[end + 1] == japchars[n] || c[end - 1] == japchars[n] || c[end] == japchars[n]) {
								separatable = false;
							}
						}`
